contracts/aggregators/light-client/src/contract.rs

```diff
@@ -0,0 +1,133 @@
+const CACHE_SIZE: usize = 6;
+
+/// sudo is the main entrypoint for the contract. It can only be called by modules.
+/// This function handles:
+///  * the deserialization of the input msg
+///  * aggregation over the VE light client inputs
+///  * updating contract state to store agreed upon state updates
+#[cfg_attr(not(feature = "library"), entry_point)]
+pub fn sudo(_: DepsMut, _: Env, _: SudoMsg) -> ContractResult<Response> {
+    ContractResult::Ok(Response::new())
+}
+
+/// instantiate is used to construct the contract
+#[cfg_attr(not(feature = "library"), entry_point)]
+pub fn instantiate(
+    deps: DepsMut,
+    _env: Env,
+    info: MessageInfo,
+    _: InstantiateMsg,
+) -> Result<Response, ContractError> {
+    set_contract_version(deps.storage, CONTRACT_NAME, CONTRACT_VERSION)?;
+    Ok(Response::new()
+        .add_attribute("method", "instantiate")
+        .add_attribute("owner", info.sender))
+}
+
+pub mod execute {
+    use super::*;
+    use crate::state::ChainHashes;
+
+    /// write_merkle_roots implements the state update method of the contract.
+    /// Merkle roots are input using a map of chain_id to hash value.
+    /// If a chain has reached the maximum cache size, it evicts the oldest entry and
+    /// inserts a new one.
+    /// Otherwise, it writes a new vector to state for the chain.
+    pub fn write_merkle_roots(
+        deps: DepsMut,
+        merkle_roots: Vec<(String, Binary)>,
+    ) -> Result<Response, ContractError> {
+        for (chain_id, merkle_hash) in merkle_roots.iter() {
+            // Get the existing vector of merkle roots for the chain_id
+            // let mut root_set: Vec<Binary>;
+            let mut root_set: ChainHashes;
+            if MERKLE_ROOTS.has(deps.storage, chain_id.to_string()) {
+                root_set = MERKLE_ROOTS.load(deps.storage, chain_id.clone()).unwrap();
+                if root_set.hashes.len() == root_set.max_size {
+                    root_set.hashes.remove(0);
+                }
+                root_set.hashes.push(merkle_hash.clone());
+            } else {
+                root_set = ChainHashes {
+                    chain_id: chain_id.clone(),
+                    hashes: Vec::new(),
+                    max_size: CACHE_SIZE,
+                };
+                root_set.hashes.push(merkle_hash.clone());
+            }
+            MERKLE_ROOTS.save(deps.storage, chain_id.clone(), &root_set)?;
+        }
+        Ok(Response::new())
+    }
+}
+
+#[cfg_attr(not(feature = "library"), entry_point)]
+pub fn query(deps: Deps, _env: Env, msg: QueryMsg) -> StdResult<Binary> {
+    match msg {
+        QueryMsg::LookupHash { chain_id, hash } => {
+            to_json_binary(&query::lookup_hash(deps, chain_id, hash)?)
+        }
+    }
+}
+
+pub mod query {
+    use super::*;
+    use crate::msg::LookupHashResponse;
+    use cosmwasm_std::StdError;
+
+    pub fn lookup_hash(
+        deps: Deps,
+        chain_id: String,
+        hash: Binary,
```

Comment:
wont the merkle root hash itself just be a string?

Reply:
I thought it would be an array of bytes--i.e. a binary?